src/index.ts

```diff
@@ -1,52 +1,76 @@
-import { NameHash } from './types/NameHash';
-import { SecondLevelDomain } from './types/SecondLevelDomain';
-import { TopLevelDomain } from './types/TopLevelDomain';
-
-export const TEST_TLD_TOPIC_ID = '0.0.48097305';
-export const MAIN_TLD_TOPIC_ID = '0.0.1234189';
-
+import { hashDomain } from "./hashDomain";
+import { MemoryCache } from "./MemoryCache";
+import { MirrorNode, NetworkType } from "./mirrorNode";
+import { PollingTopicSubscriber } from "./topicSubscriber/pollingTopicSubscriber";
+import { ICache } from "./types/Cache";
+import { NameHash } from "./types/NameHash";
+import { SecondLevelDomain } from "./types/SecondLevelDomain";
+import { TopLevelDomain } from "./types/TopLevelDomain";
+
+export const TEST_TLD_TOPIC_ID = "0.0.48097305";
+export const MAIN_TLD_TOPIC_ID = "0.0.1234189";
+
+export {
+  ICache,
+  Links,
+  MessageObject,
+  MessagesResponse,
+  NFT,
+  NFTsResponse,
+  NameHash,
+  SecondLevelDomain,
+  TopLevelDomain,
+} from "./types";
 
 export class Resolver {
   mirrorNode: MirrorNode;
-
   private _isCaughtUpWithTopic = new Map<string, boolean>();
   private _subscriptions: (() => void)[] = [];
+  private cache: ICache;
 
-  constructor(networkType: NetworkType, authKey = '') {
+  constructor(networkType: NetworkType, authKey = "", cache?: ICache) {
     this.mirrorNode = new MirrorNode(networkType, authKey);
+    if (!cache) {
+      this.cache = new MemoryCache();
+    } else {
+      this.cache = cache;
+    }
   }
 
   /**
    * @description Initializes all topic subscriptions.
    */
   public init() {
     this.getTopLevelDomains().then(() => {
-      const knownTlds = Array.from(domainCache.tld.values());
-      knownTlds.forEach(tld => {
-        this.getSecondLevelDomains(tld.topicId);
-      })
+      this.cache.getTlds().then((knownTlds) => {
+        if (knownTlds) {
+          knownTlds.forEach((tld) => {
+            this.getSecondLevelDomains(tld.topicId);
+          });
+        }
+      });
     });
   }
 
   public async dispose() {
-    await Promise.all(this._subscriptions.map(unsub => unsub()));
+    await Promise.all(this._subscriptions.map((unsub) => unsub()));
   }
 
   /**
    * @description Resolves a Second Level Domain to the wallet address of the domain's owner
    * @param domain: {string} The domain to query
    * @returns {Promise<AccountId>}
    */
-  public async resolveSLD(domain: string): Promise<string> {
+  public async resolveSLD(domain: string): Promise<string | undefined> {
     const nameHash = hashDomain(domain);
     const sld = await this.getSecondLevelDomain(nameHash);
-    const [tokenId, serial] = sld.nftId.split(':');
-    const nft = await this.mirrorNode.getNFT(tokenId, serial);
-    return nft.account_id;
+    if (sld) {
+      const [tokenId, serial] = sld.nftId.split(":");
+      const nft = await this.mirrorNode.getNFT(tokenId, serial);
+      return nft.account_id;
+    } else {
+      return Promise.resolve(undefined);
+    }
```

Comment:
same concept as before I believe we should enter an error state here as the consumer will have an easier time knowing what is happening